src/data/simple_dmatrix.h

```diff
@@ -26,6 +28,118 @@ class SimpleDMatrix : public DMatrix {
   explicit SimpleDMatrix(std::unique_ptr<DataSource<SparsePage>>&& source)
       : source_(std::move(source)) {}
 
+  template <typename AdapterT>
+  explicit SimpleDMatrix(AdapterT* adapter, float missing, int nthread) {
+    // Set number of threads but keep old value so we can reset it after
+    const int nthreadmax = omp_get_max_threads();
+    if (nthread <= 0) nthread = nthreadmax;
+    int nthread_original = omp_get_max_threads();
+    omp_set_num_threads(nthread);
+
+    source_.reset(new SimpleCSRSource());
+    SimpleCSRSource& mat = *reinterpret_cast<SimpleCSRSource*>(source_.get());
+    std::vector<uint64_t> qids;
+    uint64_t default_max = std::numeric_limits<uint64_t>::max();
+    uint64_t last_group_id = default_max;
+    bst_uint group_size = 0;
+    auto& offset_vec = mat.page_.offset.HostVector();
+    auto& data_vec = mat.page_.data.HostVector();
+    uint64_t inferred_num_columns = 0;
+
+    adapter->BeforeFirst();
+    // Iterate over batches of input data
+    while (adapter->Next()) {
```

Comment:
I'm not sure why you want to move the data handling logic into SimpleDMatrix.  Unless you want to merge `*Source` and `*DMatrix`, there is a (vague) line drawn between them.  The source handles raw data as backend class, while DMatrix is a frontend class that provides user interface for manipulating processed data, like transforming from CSR to CSC and vice versa.  I sometime visualize a workflow to understand the code, previously I thought it should be:
```
numpy/cudf  -> c_api -> (*)adapter -> (*)source -> (*)dmatrix   # (*) means various subclasses
```
But if the function is placed in `SimpleDmatrix`, this workflow doesn't apply.  This kind of visualization is rather important to me as I can't remember the code, but understanding the workflow is easy.

Reply:
I do want to merge them. They are so closely tied that i dont think we need both. I want to have self contained constructors for the dmatrix.